ocaml/xapi/xapi_clustering.ml

```diff
@@ -127,24 +127,26 @@ let find_cluster_host ~__context ~host =
     raise Api_errors.(Server_error(internal_error, [msg; (Ref.string_of host)]))
   | _ -> None
 
-let get_master_pif ~__context =
-  match find_cluster_host ~__context ~host:Helpers.(get_master ~__context) with
-  | Some self -> Db.Cluster_host.get_PIF ~__context ~self
-  | None -> raise Api_errors.(Server_error (internal_error, [ "No cluster_host exists on master" ]))
-
+(** Best-effort attempt to find a network common to the entire cluster *)
 let get_network_internal ~__context ~self =
-  let cluster_network = Db.PIF.get_network ~__context ~self:(get_master_pif ~__context) in
-  if List.exists
-    (fun cluster_host ->
-      let pif = Db.Cluster_host.get_PIF ~__context ~self:cluster_host in
-      let cluster_host_network = Db.PIF.get_network ~__context ~self:pif in
-      cluster_host_network <> cluster_network
-    ) (Db.Cluster_host.get_all ~__context)
-  then
-    debug "Not all cluster hosts of cluster %s on same network %s"
-      (Ref.string_of self) (Ref.string_of cluster_network);
-
-  cluster_network
+  let get_cluster_host_network self =
+    Db.Cluster_host.get_PIF ~__context ~self
+    |> (fun self -> Db.PIF.get_network ~__context ~self)
+  in
+  let common network = List.for_all (fun self -> network = (get_cluster_host_network self)) in
+  begin match find_cluster_host ~__context ~host:(Helpers.get_master ~__context) with
+  | Some cluster_host ->
+    get_cluster_host_network cluster_host
+    (* If cluster_host exists on master, we simply return the network of its PIF *)
```

Comment:
Maybe we could add a warning here like before, if not all cluster hosts are on the same network. Because I assume the cluster hosts should always be on the same network, whether or not the master is in the cluster.
And if this is true, I think we should fail in both cases - that's the best behaviour, because we call this when we add a cluster host. It's better not to add a cluster host to a cluster that inconsistent and falling apart because members are on different networks. This would also make the behaviour of this function more consistent - we return a common network if there is one , otherwise we fail.